widget/embedded/src/utils/common.ts

```diff
@@ -81,3 +83,40 @@ export const generateRangeColors = (name: string, color: string) => {
 
 export const getContainer = () =>
   document.getElementById('swap-box') as HTMLElement;
+
+export const generateColors = (
+  mainColors: { [x: string]: string },
+  colors?: WidgetColors
+) => {
+  if (!colors || !Object.entries(colors).length) {
+    return {};
+  }
+  let changeColors = false;
+  let listOfColors = { ...mainColors };
+  for (const color in colors) {
+    if (
+      !!colors[color as Colors] &&
+      colors[color as Colors] !== mainColors[color]
+    ) {
+      changeColors = true;
+      if (
+        (color as Colors) === 'background' ||
+        (color as Colors) === 'foreground'
+      ) {
+        listOfColors = {
+          ...listOfColors,
+          [color]: colors[color as Colors] as string,
+        };
+      } else {
+        listOfColors = {
+          ...listOfColors,
+          ...generateRangeColors(color, colors[color as Colors] as string),
+        };
+      }
+    }
+  }
+  if (changeColors) {
+    return listOfColors;
+  }
+  return {};
```

Comment:
You could define some variables to avoid duplicate codes. like this:

```
let changeColors = false;
let listOfColors = { ...mainColors };

for (const colorKey in colors) {
  const color = colors[colorKey as Colors];

  if (color && color !== mainColors[colorKey]) {
    changeColors = true;

    if (colorKey === 'background' || colorKey === 'foreground') {
      listOfColors = {
        ...listOfColors,
        [colorKey]: color,
      };
    } else {
      listOfColors = {
        ...listOfColors,
        ...generateRangeColors(colorKey, color),
      };
    }
  }
}

return changeColors ? listOfColors : {};
```

Reply:
fixed